### ISSUE TITLE
when trying load rewarded video  second time

### ISSUE THREAD

**Issue by sashastg** [mopub/mopub-ios-sdk#356]
One line summary of the issue here

- [x] I am submitting a bug report for existing functionality
- [x] I visited https://developers.mopub.com/ and found no answer
- [x] I checked https://twittercommunity.com/c/advertiser-api/mopub to make sure that this issue has not already been filed
- [x] I checked to make sure that this issue has not already been filed

#### MoPub SDK Version:
5.15.0
#### Device model and OS Version:
iphone x, 14.2   (all devices same)
#### Xcode Version:
12.2
#### Ad Unit IDs used in reproducing the issue:
443c215fd0e44ef597ab5c334e5ee69b

#### Steps to reproduce the behavior:
I attached link to video with issue. Please check it

#### Expected behavior:
I expected to get new rewarded video.

#### Observed behavior:
As you can see on video, i have issue when i am trying to load rewarded video second time.
I got immediately event  rewardedVideoAdDidLoad, but after when i call method:
MPRewardedVideo.hasAdAvailable -  i got false
Also if i am trying to: MPRewardedVideo.presentAd    nothing happens.

NOTE!  This happens only when marketplace is off.
And only with second video.  (With 1st  all goes well).
Please check links below.

#### Evidence:

video of issue:
https://drive.google.com/file/d/1aOXN2tGQvtqASA7y2dEHaWzjslXNHGht/view?usp=sharing

1 file code:
https://drive.google.com/file/d/1Hk15ZVwB85xIxWdp6qSzpDpyYysxiID9/view?usp=sharing

project with issue:
https://drive.google.com/file/d/1zns46deUHOEsCS6gtMYMa_slxUW1Zm4l/view?usp=sharing

**Comment by chauduyphanvu** [comment]
@sashastg, your project works for me and does not reproduce that issue. I can load and show multiple videos one after the other in the manner you used. Here's a screen recording of what I did: https://chauduyphanvu.s3.us-east-2.amazonaws.com/demos/mopub-ios-sdk-issues-356.mov.

Let us know if the issue persists. You may also reference our iOS sample app on this repository if needed.
Vu Chau
MoPub Team

**Comment by chauduyphanvu** [comment]
@sashastg, I was working on another issue today that might help with this issue. Make sure you run `pod update` to get latest adapter versions (that's what I did, which explained why I did not reproduce the issue in your sample app). Old adapter versions will not work properly with MoPub 5.15.0.

Since you are using Chartboost, refer to the [Chartboost changelog](https://github.com/mopub/mopub-ios-mediation/blob/master/Chartboost/CHANGELOG.md) for the latest adapter versions. You want to be on v8.3.1.1.1 or newer if you are using MoPub 5.15.0.

**Comment by KevinTydlacka** [comment]
I have an open support ticket too but want to document here for anyone else seeing this issue.

I see this issue on 5.15.0 using a blank/empty iOS application, integrating MoPub manually, using only MoPub Marketplace ads (MP Test Ad Unit IDs). This is unrelated to 3rd party network ads for me.

After initializing MP, loading an interstitial works fine the first time. But if you request a new interstitial in the `interstitialDidDisappear` delegate callback you get this behavior:

```
2020-12-11 15:51:05.908486-0600 mopubInterstitialDebug[83733:9774779] mopubInterstitialDebug: btnShowTapped
2020-12-11 15:51:05.908705-0600 mopubInterstitialDebug[83733:9774779] mopubInterstitialDebug: Interstitial is ready
2020-12-11 15:51:05.909002-0600 mopubInterstitialDebug[83733:9775150]
    [MoPub][MPInterstitialAdManager][4f117153f5c24fa6a3a92b818a5eb630] Attempting to show ad
2020-12-11 15:51:05.909160-0600 mopubInterstitialDebug[83733:9775150]
    [MoPub][MPMoPubFullscreenAdAdapter] Adapter MPMoPubFullscreenAdAdapter attempting to show ad
2020-12-11 15:51:05.913169-0600 mopubInterstitialDebug[83733:9775150]
    [MoPub][MPInterstitialAdManager][4f117153f5c24fa6a3a92b818a5eb630] Ad will appear
2020-12-11 15:51:06.429362-0600 mopubInterstitialDebug[83733:9775565]
    [MoPub][MPAnalyticsTracker] Tracking impression: https://ads.mopub.com/m/imp?cid=<redacted for public GitHub post>
2020-12-11 15:51:06.430651-0600 mopubInterstitialDebug[83733:9775150]
    [MoPub][MPOpenMeasurementTracker] Viewability tracker (0x2814b65a0) ad event: impression
2020-12-11 15:51:06.430826-0600 mopubInterstitialDebug[83733:9774779] mopubInterstitialDebug: InterstitialDelegate - interstitialDidAppear for MPID: 4f117153f5c24fa6a3a92b818a5eb630
2020-12-11 15:51:06.430851-0600 mopubInterstitialDebug[83733:9775150]
    [MoPub][MPInterstitialAdManager][4f117153f5c24fa6a3a92b818a5eb630] Ad did appear
2020-12-11 15:51:06.431300-0600 mopubInterstitialDebug[83733:9775566]
    [MoPub][MPMoPubFullscreenAdAdapter] Adapter MPMoPubFullscreenAdAdapter successfully showed ad
2020-12-11 15:51:11.685676-0600 mopubInterstitialDebug[83733:9775566]
    [MoPub][MPOpenMeasurementTracker] Viewability tracker (0x2814b65a0) stopped session
2020-12-11 15:51:11.686926-0600 mopubInterstitialDebug[83733:9774779] mopubInterstitialDebug: InterstitialDelegate - interstitialWillDisappear for MPID: 4f117153f5c24fa6a3a92b818a5eb630
2020-12-11 15:51:11.686958-0600 mopubInterstitialDebug[83733:9775565]
    [MoPub][MPInterstitialAdManager][4f117153f5c24fa6a3a92b818a5eb630] Ad will disappear
2020-12-11 15:51:12.209102-0600 mopubInterstitialDebug[83733:9774779] mopubInterstitialDebug: InterstitialDelegate - interstitialDidDisappear for MPID: 4f117153f5c24fa6a3a92b818a5eb630
2020-12-11 15:51:12.209138-0600 mopubInterstitialDebug[83733:9775566]
    [MoPub][MPInterstitialAdManager][4f117153f5c24fa6a3a92b818a5eb630] Ad did disappear
2020-12-11 15:51:12.209300-0600 mopubInterstitialDebug[83733:9774779] mopubInterstitialDebug: loadInterstitial
2020-12-11 15:51:12.209589-0600 mopubInterstitialDebug[83733:9774779] mopubInterstitialDebug: InterstitialDelegate - interstitialDidLoadAd for MPID: 4f117153f5c24fa6a3a92b818a5eb630
2020-12-11 15:51:12.209627-0600 mopubInterstitialDebug[83733:9775566]
    [MoPub][MPInterstitialAdManager][4f117153f5c24fa6a3a92b818a5eb630] Attempting to load ad
2020-12-11 15:51:13.305003-0600 mopubInterstitialDebug[83733:9775150]
    [MoPub][MPOpenMeasurementTracker] Viewability tracker (0x2814b65a0) deallocated
2020-12-11 15:51:21.915493-0600 mopubInterstitialDebug[83733:9774779] mopubInterstitialDebug: btnShowTapped
2020-12-11 15:51:21.915722-0600 mopubInterstitialDebug[83733:9774779] mopubInterstitialDebug: Interstitial is NOT ready
```

`interstitialDidLoadAd` fires immediately after requesting the ad, which is suspicious, and a subsequent attempt to show it shows that the ad is, in fact, not ready.

Temporary workaround is to delay requesting a new ad (1 second works for me). However, this was not required in previous SDK versions so I'm viewing this as a bug. Requesting a new ad upon the dismissal of one is a reasonable and common use case. See my support ticket to download and run the test project I made showcasing this bug.

**Comment by chintan100** [comment]
> `interstitialDidLoadAd` fires immediately after requesting the ad, which is suspicious, and a subsequent attempt to show it shows that the ad is, in fact, not ready.

I agree with @KevinTydlacka as I'm facing exactly the same issue. An older project with `MoPub SDK 5.14.1` works correctly but the latest `MoPub SDK 5.15.0` does have this issue. 

As a temporary workaround, I have added a 2 seconds delay for reloading the ad and it works fine. 

```
- (void)interstitialDidDisappear:(MPInterstitialAdController *)interstitial{
    
    dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(2.0 * NSEC_PER_SEC)), dispatch_get_main_queue(), ^{
        
        [self loadInterstitialAd];
    });
}
```

**Comment by chauduyphanvu** [comment]
@KevinTydlacka - could you share a repro project with us?
@chintan100 - is this happening with MoPub Marketplace or mediated ads?

**Comment by chintan100** [comment]
Happening with MoPub test interstitial id: `4f117153f5c24fa6a3a92b818a5eb630`. So I think its a MoPub SDK issue and not caused by any 3rd party SDKs.

**Comment by chauduyphanvu** [comment]
Thanks for the report. We were able to reproduce this issue too and will be investigating it. Stay tuned!

**Comment by KevinTydlacka** [comment]
Sure: https://drive.google.com/file/d/1itktEMBGFmBcZLQBLKdDASwkkAQlepJy/view?usp=sharing

I also sent it over in response to support ticket 00218272

**Comment by KevinTydlacka** [comment]
@chauduyphanvu can you give a (rough) timeline on when you guys might push a fix for this? We’re waiting to update all our apps with the next MP SDK version and are waiting on this fix. Was hoping to get the work done before Christmas.

**Comment by chauduyphanvu** [comment]
> @chauduyphanvu can you give a (rough) timeline on when you guys might push a fix for this? We’re waiting to update all our apps with the next MP SDK version and are waiting on this fix. Was hoping to get the work done before Christmas.

The current ETA is early 2021. We do not have more releases scheduled for the rest of this year. I'll post an update once the upcoming SDK is out!

**Comment by JZWDream** [comment]
I also encountered the same problem. My solution is to write a 0.1s delay in loading the next ad in the delegate that closes the ad, and the ad can be preloaded and played successfully.

```
- (void)rewardedVideoAdDidDisappearForAdUnitID:(NSString *)adUnitID {
     dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(0.1 * NSEC_PER_SEC)), dispatch_get_main_queue(), ^{
            // TODO:load ad
        });
}
```

**Comment by chauduyphanvu** [comment]
Hi all, this issue has been addressed in the [5.16.0 SDK release](https://github.com/mopub/mopub-ios-sdk/releases/tag/5.16.0). Let us know if you have any questions!

In the process, we made some callback changes in the SDK. Refer to our [dev docs](https://developers.mopub.com/publishers/ios/rewarded-ad/#step-7-optionally-implement-lifecycle-callbacks) and the [API Reference](https://developers.mopub.com/publishers/reference/ios/MPRewardedAdsDelegate/) for the updated list of callbacks.
